README
------------------------------------------------------------------------------------------
File		:	README 
Project	:	COMPATIBLE ONE ACCORDS
Author	:	Iain James Marshall
Created	:	19th November 2011
Updated	:	29th November 2011
------------------------------------------------------------------------------------------

INTRODUCTION 
------------
This file provides useful information about the production, installation, behaviour and use 
of the CompatibleOne Accords Brokering and Provisioning Platform. It is divided into three
primary sections covering production, installation and operation. Additional sections then
following providing background information pertaining to Security, Authentication, Monitoring,
Data and Temporary Files. 

PRODUCTION
----------
If you are reading this file then you will have cloned the accords-platform from the 
CompatibleOne OW2 Gitorious and will perhaps want to build the binary package.

Currently two build systems are provided. The first one for debian based systems is
based on the standard Debian packaging system. Make sure you have 'dpkg-dev' installed and
type 'dpkg-buildpackage -rsudo -b' in this directory.
The other one, the build.linux script, is to be used for non Debian based Linux platforms.

In both cases, to build the Accords Platform Binary package, simply execute the required
build script and wait till the compilation and link has all completed.

The Accords platforms depend on the following other tools and libraries which need to be 
installed on your platform before launching your build script.

	AUTOCONF
	
        AUTOMAKE
	
        AUTOTOOLS
	
        PYTHON-DEV
	
        PYTHON-BOTO

        MAKE

	GCC

	OPENSSL

	UUID

If you get this package from git repository, you must run:
# ./autogen.sh

To build this package, you can run:
# ./configure [options] (see ./configure --help for options)
# make

PLATFORM SPECIFICS
------------------

On some compilers (gcc 4.4), some additional flags are needed for the compiler.
Try a compbination of:
CFLAGS="-Wl,--no-as-needed"
LDFLAGS="-ldl"

You can pass these flags to configure script, or to make invocation.

INSTALLATION
------------

To install all files, run (as root):
# make install


PYACCORDSSDK CONFIGURATION
--------------------------

To configure the PyAccordsSDK path, run the following (as root):
# ./pyaccords_config.sh


OPERATION
---------

If you have not installed files, you must run the following commands from 'scripts' dir.

ACCORDS package provides a configuration script called 'accords-config'.

The configuration script may take several parameters which will influence the way in which the
installation will be performed.

1) 	configuration file name
2) 	host name of the platform
3)	base port number of the installation
4)	publisher url of the form http://host:port

5)	this parameter when present will activate security
6)	this parameter when present will specify the monitoring options

When no parameters are passed the installation will be performed as if it had been launched as:

	accords-config accords.xml http://127.0.0.1 8086 http://127.0.0.1:8086

The configuration procedure will create the master configuration file for all of the components 
of the Accords platform and when security is active will generate the security configurations,
certificates and private keys for all the OCCI server modules comprising the system. In addition,
a collection of command scripts will be generated that are configured facilitating the use of 
the different components of the platform.

The following collection of command scripts are to be used for the operation of the platform. 

1)	co-start
	This script must be launched to start the Accords Platform and bring the servers online and 
	takes no command line parameters. The script launchs the collection of run scripts, one for
	each of the accords platform component servers. Each server will be launched in background
	operation mode as a daemon process.

		co-start

2)	co-status
	This script may be launched to inspect the status of the Accords Platform. The status inspection 
	operation may be performed by hand using either the ps command and grepping for the processes 
	which use the the accords.xml configuration file. It may also be performed using netstat to detect
	the individual server's listening socket.

		co-status

3)	co-parser
	This script allows a manifest to be submitted for parsing and results in the creation
	of a fully qualified plan. The script takes a single parameter, the stub name of the xml
	manifest file without the xml file extension. The script launchs the executable bin/testcp
	and defines the required publisher and security information as described during installation.

		co-parser platform

4)	co-broker
	This script allows the resources described by a plan to be provisioned as a service and started.
	The script takes a single parameter, the stub name of the xml plan file without the xml file 
	extension. The script launchs the executable bin/testcb and defines the required publisher 
	and security information as described during installation.

		co-broker platform

5)	co-resolver
	This script allows the publisher to be queried for category service provider information.
	The script takes any quantity of command line parameters, expected to be category names
	that will be requested of the publisher for resolution of their service provider identity.
	The recent version of the executable bin/testresolver allows a command line option --price
	allowing activation of resolution of price information for each category processed.

		co-resolver compute storage network system package

		co-resolver --price manifest service contract openstack opennebula

6)	co-command
	This script allows a provisioned service to be started, stopped or saved. The script expects
	any number of command line parameters, the first of which must be one of the two keywords 
	"start" or "stop". The subsequent parameters are to be valid service instance description
	filenames in the service directory. The corresponding command will be performed on each of the 
	service instance descriptions bringing all contracts of the service up or down as required.

		co-command start service/uuid-of-service-instance

		co-command stop service/uuid-of-service-instance

7)	co-stop
	This script must be launched to stop the Accords Platform and bring the servers offline.
	This script takes no command line parameter and will perform a killall -15 operation for
	each of the service components of the accords platform. Service components intercept this
	signal and perform a garceful exit after releasing all ressources and saving any data.

		co-stop

Several additional binary executable test programs are offered by the package allowing access to the 
different cloud provider platforms currently supported for use by the Accords Platform. These tools 
allow inspection of the different resources of the platforms and have a platform specific command 
line syntax.

1)	testos
	This program provides a test client interface for the OpenStack platform and although it was
	initially aligned for operation with the version 1.0 since the arrival of Diablo and version 1.1
	interoperability has not been maintained by the platform itself. The OpenStack platform offers
	several resource types: flavors, images, servers and metadata. The specific command line syntax 
	for the management of resources using this tool may be consulted in the integrated help screen
	that may be obtained by simply launching the tool with no command line parameters.
	
2)	teston

3)	testaz

SECURITY
--------
During the installation operation if security has been requested then a security directoy will be 
created into which the security configuration files of each of the platform components will be 
generated. In addition the private key and certificates will also be generated into this directory.

The files generated into the security directory are not to be world readable for obvious reasons.

Each individual security configuration file defines the private key, certificate, user name, password 
and the security options of each of the server components. 

The Private Keys and Certificates are currently only self signed and will therefore cause Incorrect 
Certificate alerts and errors to be thrown when used with the Certificate Exchange and Validation 
Mechanisms activated. This highlights the need for a mechanism within the CompatibleOne platform for 
the production and delivery of Valid Certification Authority Signed certificates for use not only by 
the Accords platform components but also by the provisioned resources and products delivered by the 
platform. 

The security option values are defined as follows:

1,2,4		These three option values are undefined in this SSL / TLS implementation for ACCORDS.

8		REQUEST PEER CERTIFICATE
		When this option value is set then the server will issue a Send Certificate request to
		the client during the Server Hello phase of handshake and cypher negociation. 
		This option instructs the Client to validate the Certificate chain received from the 
		Server prior to sending the Client side certificate if one is available. Self Signed
		Certificates with throw "Incorrect Certificate" alerts when this option is active.
		If a Certificate is not exchanged then no failure will occur since it not mandatory.

16		REQUIRE PEER CERTIFICATE
		This options requires an exchange of valid Certificates by both Client and Servers.
		An error will occur if no Certificate is made available or if the Certificate is not
		Correct.

32		DER KEY FILE FORMAT
		When this option is set it indicates that the Private Key file is not a base64 PEM
		Ascii encoded file but rather a binary DSR,ASN or otherwise encoded file.

64		DER CERTIFICATE FILE FORMAT
		When this option is set it indicates that the Certificate File is not a base64 PEM
		Ascii encoded file but rather a binary DSR,ASN or otherwise encoded file.

128		SSL COMPATIBLE
		when this option is set then the SSL compatible mode will be activated and as such the
		client connection handshake will commence with an SSL V2 or V3 Hello Message. This may
		be accepted by SSL V1, V2, V3 and TLS v1 servers. However an RFC dating from April 2011
		is now recommending that this is bad practice since it allows MTM attacks to influence
		both the client and server to be misled into believing that the other party requires an
		inferior protocole. The RFC is strongly advocating that this v23 Client Hello should be
		phased out as quickly as possible. In compliance with this recommendation the TLS / SSL 
		functions of the Accords Platform will issue a TLS V1 Client Hello Message and as a
		server will only communicate with Clients behaving accordingly. This requirement may be
		relaxed by use of this option value.

256		INTERNAL SSL CERTIFICATE VERIFICATION
		This option may be set when Exchange of Self-Signed Certificates by Clients and Servers
		is required. This is necessary for instance when working with Windows Azure platforms
		which expressly request provision of a self signed certificate by the client endpoint.

These options are necessary during the development and test phases of the Accords platform and should 
be used carefully and only by persons with a solid understanding of the RFC for TLS and their implications 
towards the resulting platform security. 

HINT
----
The value of the security options is provided by two different
variables "options" and "optionst" inside the "install.linux" and "install.debian" scripts to allow the 
value to be set for servers and for tools independantly. It is however recommened that they be set to the
same value to ensure homogenous operation.

AUTHENTICATION
--------------
When security has been requested and activated during the installation phase, an individual user name 
and associated password will have been generated to the security configuration file of each component.
The user category management service database will be generated to include the collection of valid user 
name and password authentication credentials. The corresponding authorisation option of the security 
configuration file, when set, will cause the service component to present the credentials to COSS for
authentication prior to requesting the creation of an authorization token allowing collaboration by the 
service in the Accords Platform. This authorization credential will be presented in the OCCI/REST/HTTP 
header of ALL Request and Response messages sent by the Client of Server endpoints of the Service component.
For obvious reasons neither the PUBLISHER nor the COSS modules are required to perform authentication and
authorization though this will be introduced as a retro fit requirement at a later date.
 
MONITORING
----------
During the platform operation, log and trace information will be generated as described by the
monitoring parameter of the installation script.

0)	No monitoring
1) 	All monitoring is sent to the file co-log
2)	All monitoring is sent to the Module COMONS, except for the PUBLISHER, COMONS and COSS
3)	Monitoring information is sent to both the file co-log and to COMONS
6)	Optimised Monitoring is sent to the Module COMONS, except for the PUBLISHER, COMONS and COSS
	and to the file co-log.

When COMONS monitoring is active then only the service modules started after COMONS are able to make 
use of the centralised monitoring service. Therefore the activation of CO-LOG monitoring is necessary
for the fundamental modules of the platform : PUBLISHER, COSS and COMONS. 

DATA
----
The collection of XML files cords_*.xml and links_*.xml represent the internal state of the different
server components of the platform. They are updated as their internal state changes and will be reloaded
when the corrresponding component restarts.

TEMPORARY FILES
---------------
The directory "rest/" contains the contents of the messages exchanged by the server components during 
operation and may be consulted. These may have several different extensions depending on the source
of the information and may be XML, JSON, OCCI, HTML or TMP files. 

HINT
----
This directory will need to be cleaned
at regular intervals in cases of prolonged operation otherwise performance degradation will occur due to
the number of files that accumulate here. 

-----------
End of File
-----------


debian/changelog
compatibleone-accords (1.0-12) unstable; urgency=low

  * Upstream update.
  * Add mysql/postgres build deps.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 27 Jan 2014 08:21:00 +0100

compatibleone-accords (1.0-11) unstable; urgency=low

  * Upstream update.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 27 Aug 2013 16:02:00 +0200

compatibleone-accords (1.0-10) unstable; urgency=low

  * Add compatibleone-oaprocci package.
  * fix to build from the new github repository.
  * move comodel to base package and also include strukt.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 06 Jun 2013 10:42:12 +0200

compatibleone-accords (1.0-9) unstable; urgency=low

  * Upstream code update (new codegen code).

 -- Vincent Renardias <vincent@renardias.com>  Mon, 06 May 2013 15:45:02 +0200

compatibleone-accords (1.0-8) unstable; urgency=low

  * Upstream code update (big dashboard scripts update).

 -- Vincent Renardias <vincent@renardias.com>  Mon, 06 May 2013 15:44:46 +0200

compatibleone-accords (1.0-7) unstable; urgency=low

  * Upstream code update.
  * New package: compatible-ec2procci.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 18 Mar 2013 15:45:01 +0100

compatibleone-accords (1.0-6) unstable; urgency=low

  * Improved Python bindings.
  * Add 'compatibleone-cnprocci' package for ComputeNext interface.
  * Add 'compatibleone-eziprocci' package for EasyClouds interface.
  * Upstream code update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 11 Mar 2013 10:34:58 +0100

compatibleone-accords (1.0-5) unstable; urgency=low

  * Upstream update.
  * Fix dashboard scripts.

 -- Vincent <vincent@renardias.com>  Thu, 21 Feb 2013 22:17:00 +0100

compatibleone-accords (1.0-4) unstable; urgency=low

  * Add new package compatibleone-dashboard (more work needed).
  * Clean-up control file, start adding the php dashboard
  * Replace hardwired emails.

 -- Vincent <vincent@renardias.com>  Thu, 21 Feb 2013 18:49:00 +0100

compatibleone-accords (1.0-3) unstable; urgency=low

  * Upstream update (tree SHA: a5c42c620e43d7a698c35e725cb7cf7fbba61181)
  * Add libcocspi0 and libcocspi-dev packages (+bump up solib deps).

 -- Vincent <vincent@renardias.com>  Wed, 13 Feb 2013 12:18:57 +0100

compatibleone-accords (1.0-2) unstable; urgency=low

  * Upstream code of jan 23rd.
  * Include selftest.csp cordscript.
  * Remove occurences of paprocci in current build.

 -- Vincent <vincent@renardias.com>  Wed, 23 Jan 2013 14:58:01 +0100

compatibleone-accords (1.0-1) unstable; urgency=low

  * Bump up version number!
  * Update solib deps accordingly.
  * Include code for the current repository.
  * Code / Comments clean-ups in debian/rules.
  * Disable python testsuite (doesn't work and most likely won't be maintained).
  * Wrote minimalist manpage for the cosacs client to shut up lintian.
  * Declare compatibleone-accords and compatibleone-accords-dev as metapackages.
  * Packages are now fully clean according to lintian (except the Python JDK).

 -- Vincent Renardias <vincent@renardias.com>  Fri, 18 Jan 2013 14:01:40 +0100

compatibleone-accords (0.20130101-1) unstable; urgency=low

  * Upstream code update + .csp sample file.
  * Set dpkg-source 1.0 format.
  * Update lib soname dep.

 -- Vincent Renardias <vincent@renardias.com>  Wed, 02 Jan 2013 10:24:36 +0100

compatibleone-accords (0.20121218-1) unstable; urgency=low

  * New upstream code + more doc + sync all binary packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 18 Dec 2012 14:43:00 +0200

compatibleone-accords (0.20121214-1) unstable; urgency=low

  * Several bug fixes regarding cosacs.

 -- Vincent Renardias <vincent@renardias.com>  Fri, 14 Dec 2012 16:58:00 +0200

compatibleone-accords (0.20121213-2) unstable; urgency=low

  * Add 'cosacs' package (client side COSACS to be run in the VM images).
  * Split 'compatibleone' in 'compatibleone-base' and 'compatibleone'.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 13 Dec 2012 17:47:00 +0200

compatibleone-accords (0.20121213-1) unstable; urgency=low

  * Major renaming of packages to be fully Debian compliant and prepare the
    integration of the Cosacs client.
  * Force lib # in the dependency and bump the version to make sure we're
    not linking against a version older than today (lib ABI change).

 -- Vincent Renardias <vincent@renardias.com>  Tue, 12 Dec 2012 15:16:00 +0200

compatibleone-accords (0.20121212-1) unstable; urgency=low

  * Upstream update.
  * Major packaging update: each lib now has its own packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 11 Dec 2012 15:47:00 +0200

compatibleone-accords (0.20121207-1) unstable; urgency=low

  * Upstream update.
  * Lots of packaging fixes (see gitorious log for the gory details)

 -- Vincent Renardias <vincent@renardias.com>  Fri, 07 Dec 2012 10:25:21 +0200

compatibleone-accords (0.20121129-1) unstable; urgency=low

  * Upstream update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 26 Nov 2012 11:43:34 +0200

compatibleone-accords (0.20121121-1) unstable; urgency=low

  * Uptream update 2012/11/21 codebase + python enabled.
  * compatibleone-example and compatibleone-testosocci are now gone.

 -- Vincent Renardias <vincent@renardias.com>  Wed, 21 Nov 2012 11:12:59 +0200

compatibleone-accords (0.20121120-1) unstable; urgency=low

  * Uptream update 2012/11/20 codebase + perm fixes.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Tue, 20 Nov 2012 15:24:44 +0200

compatibleone-accords (0.20121114-2) unstable; urgency=low

  * Uptream update 2012/11/19 codebase.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 19 Nov 2012 20:41:06 +0200

compatibleone-accords (0.20121114-1) unstable; urgency=low

  * Uptream update,
  * jpaasprocci -> paasprocci renaming.
  * fixes for java build on 64bit arch.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Wed, 14 Nov 2012 14:24:57 +0200

compatibleone-accords (0.20121022-1) unstable; urgency=low


ezvm/ezvm.c
/* -------------------------------------------------------------------- */
/*  ACCORDS PLATFORM                                                    */
/*  (C) 2011 by Iain James Marshall (Prologue) <ijm667@hotmail.com>     */
/* -------------------------------------------------------------------- */
/* Licensed under the Apache License, Version 2.0 (the "License"); 	*/
/* you may not use this file except in compliance with the License. 	*/
/* You may obtain a copy of the License at 				*/
/*  									*/
/*  http://www.apache.org/licenses/LICENSE-2.0 				*/
/*  									*/
/* Unless required by applicable law or agreed to in writing, software 	*/
/* distributed under the License is distributed on an "AS IS" BASIS, 	*/
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 	*/
/* implied. 								*/
/* See the License for the specific language governing permissions and 	*/
/* limitations under the License. 					*/
/* -------------------------------------------------------------------- */
#ifndef	_ezvm_c
#define	_ezvm_c

#include "standard.h"
#include "ezvm.h"
#include "rest.h"
#include "occi.h"
#include "document.h"
#include "cordspublic.h"
#include "occipublisher.h"
#include "occibuilder.h"
#include "cordslang.h"
#include "cp.h"

struct	accords_configuration Ezvm = {
	0,0, 
	0,0,0,0,
	(char *) 0,
	(char *) 0,
	_CORDS_DEFAULT_PUBLISHER,
	_CORDS_DEFAULT_OPERATOR,
	_CORDS_DEFAULT_USER,
	_CORDS_DEFAULT_PASSWORD,
	"http",  80,
	"xmpp",  8000,
	"domain",
	"ezvm.xml",
	"europe",
	"storage",
	"ezvm",
	(struct occi_category *) 0,
	(struct occi_category *) 0
	};

public	int	check_debug()		{	return(Ezvm.debug);		}
public	int	check_verbose()		{	return(Ezvm.verbose);		}
public	char *	default_publisher()	{	return(Ezvm.publisher);	}
public	char *	default_operator()	{	return(Ezvm.operator);	}
public	char *	default_tls()		{	return(Ezvm.tls);		}
public	char *	default_zone()		{	return(Ezvm.zone);		}

public	int	failure( int e, char * m1, char * m2 )
{
	if ( e )
	{
		printf("\n*** failure %u",e);
		if ( m1 )
			printf(" : %s",m1);
		if ( m2 )
			printf(" : %s",m2);
		printf(" **`\n");
	}
	return( e );
}

/*	---------------------------------------------	*/  
/*			e z v m _ l o a d 		*/
/*	---------------------------------------------	*/
/*	this function loads ezvm    configuration	*/
/*	from the xml configuration file.		*/
/*	---------------------------------------------	*/  
private	void	ezvm_load()
{
	load_accords_configuration( &Ezvm, "ezvm" );
	return;
}

private	int	banner()
{
	printf("\n   CompatibleOne Ezvm : Version 1.0a.0.02");
	printf("\n   Beta Version : 25/05/2012");
	printf("\n   Copyright (c) 2012 Iain James Marshall, Prologue");
	printf("\n");
	accords_configuration_options();
	printf("\n\n");
	return(0);

}

/*	------------------------------------------------------------------	*/
/*			e z v m _ i n i t i a l i s e				*/
/*	------------------------------------------------------------------	*/
private	struct rest_server * ezvm_initialise(  void * v,struct rest_server * sptr )
{
	struct	rest_extension * xptr;
	if (!( xptr = rest_add_extension( sptr ) ))
		return((struct rest_server *) 0);
	else
	{
		xptr->net = (struct connection *) 0;
		return( sptr );
	}
}

/*	------------------------------------------------------------------	*/
/*			e z v m _ a u t h o r i s e 				*/
/*	------------------------------------------------------------------	*/
private	int	ezvm_authorise(  void * v,struct rest_client * cptr, char * username, char * password )
{
	if ( strcmp( username, Ezvm.user ) )
		return(0);
	else if ( strcmp( password, Ezvm.password ) )
		return(0);
	else if (!( cptr->user = allocate_string( username ) ))
		return(0);
	else if (!( cptr->pass = allocate_string( password ) ))
		return(0);
	else	return(1);
}

/*	------------------------------------------------------------------	*/
/*			e z v m _ e x t e n s i o n 					*/
/*	------------------------------------------------------------------	*/
private	struct rest_extension * ezvm_extension( void * v,struct rest_server * sptr, struct rest_extension * xptr)
{
	return( xptr );
}

/*	------------------------------------------------------------------	*/
/*	inclusion of the actions and methods required for vm       handling	*/
/*	------------------------------------------------------------------	*/

/*	-------------------------------------------	*/
/* 	      b u i l d _ a p p l i c a t i o n  	*/
/*	-------------------------------------------	*/
private	struct rest_response * build_vm(
		struct occi_category * optr, 
		struct rest_client * cptr, 
		struct rest_request * rptr, 
		struct rest_response * aptr, 
		void * vptr )
{
	struct	cords_vm * pptr;
	struct	occi_response * zptr;
	struct	occi_response * yptr;
	char *	sptr;
	int	status;
	if (!( pptr = vptr ))
		return( rest_html_response( aptr, 400, "Failure" ) );
	else if ( pptr->state != 0 )
		return( rest_html_response( aptr, 200, "OK" ) );
	else if (!( pptr->image ))
		return( rest_html_response( aptr, 400, "incorrect image" ) );
	else if (!( pptr->provider ))
		return( rest_html_response( aptr, 400, "incorrect provider" ) );
	else if (!( pptr->application ))
		return( rest_html_response( aptr, 400, "incorrect application" ) );
	else if (!( zptr = cords_invoke_action( pptr->application, _CORDS_BUILD, _CORDS_CONTRACT_AGENT, default_tls() ) ))
		return( rest_html_response( aptr, 400, "build failure" ) );
	else if (!( zptr->response ))
	{
		zptr = occi_remove_response( zptr );
		return( rest_html_response( aptr, 400, "build failure" ) );
	}
	else if (( pptr->state = zptr->response->status) > 299 )
	{
		zptr = occi_remove_response( zptr );
		return( rest_html_response( aptr, pptr->state, "build failure" ) );
	}
	else
	{
		zptr = occi_remove_response( zptr );
		if (!( zptr = occi_simple_get( pptr->application, _CORDS_CONTRACT_AGENT, default_tls() ) ))
			return( rest_html_response( aptr, 440, "build failure" ) );
		else if (!( sptr = occi_extract_atribut( zptr, "occi", _CORDS_APPLICATION, "url" )))
		{
			zptr = occi_remove_response( zptr );
			return( rest_html_response( aptr, 441, "build failure" ) );
		}
		else if (!( pptr->url = allocate_string( sptr ) ))
		{
			zptr = occi_remove_response( zptr );
			return( rest_html_response( aptr, 500, "build failure" ) );
		}
		else	autosave_cords_vm_node(pptr);
		zptr = occi_remove_response( zptr );
		return( rest_html_response( aptr, 200, "OK" ) );
	}
}

/*	-------------------------------------------	*/
/* 	   	   c r e a t e _ v m   			*/
/*	-------------------------------------------	*/
private	int	create_cords_vm(struct occi_category * optr, void * vptr, struct rest_request * rptr)
{
	char 	buffer[2048];
	struct	occi_element *	root=(struct occi_element *) 0;
	struct	occi_element *	last=(struct occi_element *) 0;
	struct	occi_element *	eptr=(struct occi_element *) 0;
	struct	occi_response * zptr=(struct occi_response*) 0;
	char *	sptr;
	struct	occi_kind_node * nptr;
	struct	cords_vm * pptr;
	if (!( nptr = vptr ))
		return(0);
	else if (!( pptr = nptr->contents ))
		return(0);
	else if ( pptr->state )
		return(0);
	else if (!( pptr->image ))
		return(0);
	else if (!( pptr->provider ))
		return(0);
	else if (!( eptr = occi_create_element( "occi.application.image", pptr->image ) ))
		return(27);
	else if (!( eptr->previous = last ))
		root = eptr;
	else 	last->next = eptr;
	last = eptr;
	if (!( eptr = occi_create_element( "occi.application.provider", pptr->provider ) ))
		return(27);
	else if (!( eptr->previous = last ))
		root = eptr;
	else 	last->next = eptr;
	last = eptr;

	if (!( zptr = cords_create_instance( _CORDS_APPLICATION, _CORDS_CONTRACT_AGENT, root, default_tls() ) ))
		return( 0 );
	else if (!( sptr = occi_extract_location( zptr ) ))
	{
		zptr = occi_remove_response( zptr );
		return(0);
	}
	else
	{
		rest_add_http_prefix(buffer,2048,sptr);
		if ( pptr->application ) pptr->application = liberate( pptr->application );
		pptr->application = allocate_string( buffer );
		zptr = occi_remove_response( zptr );
		if ( pptr->application )
			return(0);
		else	return(27);
	}
}

/*	-------------------------------------------	*/
/* 	 	 r e t r i e v e _ v m   		*/
/*	-------------------------------------------	*/
private	int	retrieve_cords_vm(struct occi_category * optr, void * vptr, struct rest_request * rptr)
{
	struct	occi_kind_node * nptr;
	struct	cords_vm * pptr;
	if (!( nptr = vptr ))
		return(0);
	else if (!( pptr = nptr->contents ))
		return(0);
	else	return(0);
}

/*	-------------------------------------------	*/
/* 	  	    u p d a t e _ v m   		*/
/*	-------------------------------------------	*/
private	int	update_cords_vm(struct occi_category * optr, void * vptr, struct rest_request * rptr)
{
	struct	occi_kind_node * nptr;
	struct	cords_vm * pptr;
	if (!( nptr = vptr ))
		return(0);
	else if (!( pptr = nptr->contents ))
		return(0);
	else	return(0);
}

/*	-------------------------------------------	*/
/* 	  	    d e l e t e _ v m  		 	*/
/*	-------------------------------------------	*/
private	int	delete_cords_vm(struct occi_category * optr, void * vptr, struct rest_request * rptr)
{
	struct	occi_kind_node * nptr;
	struct	cords_vm * pptr;
	if (!( nptr = vptr ))
		return(0);
	else if (!( pptr = nptr->contents ))
		return(0);
	else if (!( pptr->application ))
		return(0);
	else	
	{
		occi_simple_delete( pptr->application, _CORDS_CONTRACT_AGENT, default_tls() );
		return(0);
	}
}

private	struct	occi_interface	cords_vm_interface = {
	create_cords_vm,
	retrieve_cords_vm,
	update_cords_vm,
	delete_cords_vm
	};

/*	------------------------------------------------------------------	*/
/*			e z v m _ o p e r a t i o n				*/
/*	------------------------------------------------------------------	*/
private	int	ezvm_operation( char * nptr )
{

	struct	occi_category * first=(struct occi_category *) 0;
	struct	occi_category * last=(struct occi_category *) 0;
	struct	occi_category * optr=(struct occi_category *) 0;

	set_autosave_cords_xlink_name("links_ezvm.xml");

	if (!( optr = occi_cords_vm_builder( Ezvm.domain,"vm" ) ))
		return( 27 );
	else if (!( optr->previous = last ))
		first = optr;
	else	optr->previous->next = optr;
	last = optr;

	optr->callback = &cords_vm_interface;

	if (!( optr = occi_add_action( optr,"build","",build_vm)))
		return( 28 );

	if (!( optr = occi_cords_image_builder( Ezvm.domain,"image" ) ))
		return( 27 );
	else if (!( optr->previous = last ))
		first = optr;
	else	optr->previous->next = optr;
	last = optr;

	if (!( optr = occi_cords_system_builder( Ezvm.domain,"system" ) ))
		return( 27 );
	else if (!( optr->previous = last ))
		first = optr;
	else	optr->previous->next = optr;
	last = optr;
	if (!( optr = occi_cords_package_builder( Ezvm.domain,"package" ) ))
		return( 27 );
	else if (!( optr->previous = last ))
		first = optr;
	else	optr->previous->next = optr;
	last = optr;

	rest_initialise_log( Ezvm.monitor );

	if (!( Ezvm.identity ))
		return( occi_server(  nptr, Ezvm.restport, Ezvm.tls, Ezvm.threads, first, (char *) 0 ) );
	else
	{
		initialise_occi_publisher( Ezvm.publisher, (char*) 0, (char *) 0, (char *) 0);
		return( publishing_occi_server(
			Ezvm.user, Ezvm.password,
			Ezvm.identity,  nptr, 
			Ezvm.restport, Ezvm.tls, 
			Ezvm.threads, first ) );
	}
}


/*	------------------------------------------------------------------	*/
/*				e z v m 					*/
/*	------------------------------------------------------------------	*/
private	int	ezvm(int argc, char * argv[] )
{
	int	status=0;
	int	argi=0;
	char *	aptr;
	ezvm_load();
	while ( argi < argc )
	{
		if (!( aptr = argv[++argi] ))
			break;
		else if ( *aptr == '-' )
		{
			aptr++;
			switch( *(aptr++) )
			{
			case	'v'	:
				Ezvm.verbose=1;
				continue;
			case	'd'	:
				Ezvm.debug = 0xFFFF;
				continue;
			case	'-'	:
				if (!( argi = accords_configuration_option( aptr, argi, argv )))
					break;
				else	continue;
			}
			status = 30;
			break;
		}
		else if (!( status = ezvm_operation(aptr) ))
			continue;
		else	break;
	}
	return(status);
}

/*	------------------------------------------------------------------	*/
/*					m a i n 				*/
/*	------------------------------------------------------------------	*/
public	int	main(int argc, char * argv[] )
{
	if ( argc == 1 )
		return( banner() );
	else	return( ezvm( argc, argv ) );
}


	/* --------- */
#endif	/* _ezvm_c */
	/* --------- */

